### PR TITLE
campaigns: Add basic resync UI

### PR DESCRIPTION
**Pull request by eseliger**
image

not super polished but wanted to get that out for 3.14, will follow up with improvements later when there is time

### CODE REVIEW

File: enterprise/internal/campaigns/resolvers/resolver.go [hunk 1]
```diff
@@ -629,6 +630,30 @@ func (r *Resolver) PublishChangeset(ctx context.Context, args *graphqlbackend.Pu
...
+	if err := backend.CheckCurrentUserIsSiteAdmin(ctx); err != nil {
+		return nil, errors.Wrap(err, "checking if user is admin")
+	}
+
+	changesetID, err := unmarshalChangesetID(args.Changeset)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := repoupdater.DefaultClient.EnqueueChangesetSync(ctx, []int64{changesetID}); err != nil {
```
Comment (mrnugget) [inline enterprise/internal/campaigns/resolvers/resolver.go]: I think we should check whether the Changeset exists before we enqueue it. Yes, the syncer won't break if it doesn't exist, but I don't think we should send IDs from one end to the other end of the system only to receive an error log message somewhere we didn't think of.

File: web/src/enterprise/campaigns/detail/changesets/ChangesetNode.tsx [hunk 2]
```diff
@@ -72,6 +75,16 @@ export const ChangesetNode: React.FunctionComponent<ChangesetNodeProps> = ({
             setIsLoading(false)
         }
     }
+    const enqueueChangeset: React.MouseEventHandler = async () => {
+        if (node.__typename === 'ExternalChangeset') {
+            setLastUpdatedAt(node.updatedAt)
+            await syncChangeset(node.id)
+            // campaign should be refreshed, in case the burndown chart has changed
```
Comment (mrnugget) [inline web/src/enterprise/campaigns/detail/changesets/ChangesetNode.tsx]: I'm not sure I understand the `campaignUpdates` logic here, but syncing a changeset is an asynchronous operation and you have no guarantee when it happens. I don't think it makes sense to queue a campaign update right after the enqueuing the changeset for updates.

Comment (eseliger) [reply web/src/enterprise/campaigns/detail/changesets/ChangesetNode.tsx]: you are right, that's useless 🤦‍♂

**Comment by eseliger** [comment]
@mrnugget I revisited and polished this PR up, I think it is working smooth now. This also somewhat depends on #9031

File: web/src/enterprise/campaigns/detail/changesets/ChangesetNode.tsx [hunk 3]
```diff
@@ -161,6 +187,22 @@ export const ChangesetNode: React.FunctionComponent<ChangesetNodeProps> = ({
                 )}
             </div>
             <div className="flex-shrink-0 flex-grow-0 ml-1 d-flex flex-column align-items-end">
+                {node.__typename === 'ExternalChangeset' && (
+                    <span
+                        data-tooltip={
+                            node.updatedAt === lastUpdatedAt
+                                ? 'Currently refreshing'
+                                : `Last synced ${formatDistanceToNow(
+                                      parseISO(node.updatedAt)
```
Comment (felixfbecker) [inline web/src/enterprise/campaigns/detail/changesets/ChangesetNode.tsx]: Isn't `new Date()` sufficient?
```suggestion
                                      new Date(node.updatedAt)
```

File: enterprise/internal/campaigns/resolvers/resolver.go [hunk 4]
```diff
@@ -681,6 +681,11 @@ func (r *Resolver) SyncChangeset(ctx context.Context, args *graphqlbackend.SyncC
 		return nil, err
 	}
 
+	// Check for existance of changeset so we don't swallow that error.
```
Comment (mrnugget) [inline enterprise/internal/campaigns/resolvers/resolver.go]:
```suggestion
	// Check for existence of changeset so we don't swallow that error.
```

File: web/src/enterprise/campaigns/detail/changesets/ChangesetNode.tsx [hunk 5]
```diff
@@ -76,13 +87,13 @@ export const ChangesetNode: React.FunctionComponent<ChangesetNodeProps> = ({
         }
     }
     const enqueueChangeset: React.MouseEventHandler = async () => {
+        // already enqueued
+        if (lastUpdatedAt) {
```
Comment (mrnugget) [inline web/src/enterprise/campaigns/detail/changesets/ChangesetNode.tsx]: That means you can only manually sync once per page load?

File: web/src/enterprise/campaigns/detail/changesets/ChangesetNode.tsx (hunk 3 above)
Comment (eseliger) [reply web/src/enterprise/campaigns/detail/changesets/ChangesetNode.tsx]: it only accepts proper ISO format and doesn't respect local differences like

`03-02-2020`
US: March 3, 2020
World: 3. February 2020

But it's not important here :)

File: web/src/enterprise/campaigns/detail/changesets/ChangesetNode.tsx (hunk 5 above)
Comment (eseliger) [reply web/src/enterprise/campaigns/detail/changesets/ChangesetNode.tsx]: Nope, until the updatedAt changed (ie. the sync is complete)

**Comment by eseliger** [comment]
image

WDYT?

**Pull request merged**